### ISSUE TITLE
implement a getopt command

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#3417]
The design of the fish language makes simple option parsing extremely simple, robust, and elegant. As evidenced by all the functions that ship with fish which employ patterns such as this example:

```
set -l mode
while set -q argv[1]
    switch $argv[1]
        case '-h' '--help'
            __fish_print_help abbr
            return 0
        case '-a' '--add'
            set mode add
            set needs_arg multi
    end
    set -e argv[1]
end
```

The problem is that this approach makes it impossible to handle the advanced cases, such as short option bundling, handled by the `wgetopt()` function that fish builtins use. Issue #3403 was opened by @floam to point out that the `history` function and builtin handle options differently as a consequence of not having a `getopt` command usable by fish scripts.

There is also issue #478 which intends to implement a more human friendly solution (compared to getopt) but it has been open for four years and it is unclear when it will be ready for use by fish scripts.

We can't rely on the `getopt` external command available on most (all?) UNIX distros because they differ in supported features and behavior. We currently have two functions (see *share/functions/trap.fish* and *share/functions/umask.fish*) that use the distro provided `getopt` command. Notice the hoops they have to jump through to do so. Furthermore, one of the implementations doesn't support long options. So depending on which platform you're running fish the `umask` and `trap` commands may or may not accept long options.

@alphapapa proposed we use https://github.com/fisherman/getopts. I looked at it and concluded it is no better than our current approach. It doesn't handle short option bundling. Short options have to be an ASCII letter. So you can't have a `-1` flag for example. It doesn't cleanly handle flags that take no argument, an optional argument, or a mandatory argument. You have to handle that in the case block for the option which complicates option handling and makes the logic harder to understand.

If we knew that the GNU `getopt` command was always available we would simply use it. Since it isn't I propose we implement a fish `getopt` builtin that wraps the `wgetopt()` function we use for all builtins. This should require a modest, almost trivial if you ignore writing unit tests, amount of effort to implement.

**Comment by krader1961** [comment]
There's also https://github.com/alphapapa/bucket/blob/master/getopts.fish (a pure fish implementation) but a casual review suggests it has the same limitations as the version written in awk linked to above.

**Comment by ridiculousfish** [comment]
My docopt branch and proposal still lives. My plan is to wait for C++11, find ways to reduce the complexity, and then break up the fish-side integration into reviewable pieces.

**Comment by krader1961** [comment]
> My docopt branch and proposal still lives.

I didn't doubt that. The question is whether creating a builtin `getopt` command, while suboptimal from a usability standpoint, is cheap enough to implement to make doing so useful while you continue to work on the better solution. Had someone proposed doing so two years ago I'd say the answer is quite clearly "yes". What is unclear to me today given what I saw in your docopt branch is whether the modest effort to implement `getopt` can still be justified.

Do you have a sense for whether you'll be done in the next six months, @ridiculousfish? That's rather arbitrary but I think I can implement a `wgetopt()` wrapper in under two weeks of part-time effort. Add another two to three weeks to modify our scripts to use it. So roughly one month of effort. If you're going to complete your work in less than six months then this proposal can't be justified. If docopt is going to take longer than that to see the light of day I think this proposal is justified.

**Comment by maxnordlund** [comment]
From reading #3138 it seems like the C++11 problems have been (mostly) [resolved][1]. Which should bring hope for the docopt branch. While the six months time frame may seem arbitrary, if you look back it sorta lines up with the 2.5 release (if releases happens every half year or so). So perhaps having the docopt branch merged and released sometime next spring could be a worthwhile goal, but I agree with @krader1961 that if it's not possible, then a interim solution would be nice.

[1]: https://github.com/fish-shell/fish-shell/issues/3138#issuecomment-241664748

**Comment by floam** [comment]
An emoji on says so much: my thumbs up to @krader1961 and @ridiculousfish's comments above are to indicate support for finishing docopt *sooner than later* as the cure-all and much-preferable alternative to getopt.

**Comment by floam** [comment]
What are the key differences between a c++11 fish docopt implementation and the existing c++11 docopt?

(https://github.com/docopt/docopt.cpp)

**Comment by ridiculousfish** [comment]
@floam the differences are spelled out at https://github.com/ridiculousfish/docopt_fish/

The official C++ docopt is (was?) a line-by-line port from Python.

**Comment by ridiculousfish** [comment]
@krader1961 I think six months is quite doable, yes.

**Comment by krader1961** [comment]
Okay, closing this rather than waste time on something that will be obsolete in the near future.

**Comment by hrvoj3e** [comment]
For future "googlers" this has been implemented in Fish 2.7.0 as builtin `argparse`
> https://fishshell.com/docs/current/commands.html#argparse
